wireguard/wireguard.go

```diff
@@ -0,0 +1,1504 @@
+	// 5. Simultaneous updates of wireguard, routes and rules.
+	var conflictingKeys = set.New()
+	wireguardPeerDelete := w.handlePeerAndRouteDeletionFromPeerUpdates(conflictingKeys)
+	w.updateCacheFromPeerUpdates(conflictingKeys)
+	w.updateRouteTableFromPeerUpdates()
+
+	defer func() {
+		// Flag the programmed state to be the same as the expected state for each peer. We do this even if we failed to
+		// apply the update because the routetable processing also uses this to maintain details about whether or not it
+		// has routed to wireguard. In the event of a failed update or wireguard config, a full resync will be performed
+		// next iteration which ignores the programmedInWireguard flag.
+		if len(w.peerUpdates) > 0 {
+			for name, node := range w.peers {
+				if w.shouldProgramWireguardPeer(name, node) {
+					w.logCxt.Debugf("Flag node %s as programmed", name)
+					node.programmedInWireguard = true
+				} else {
+					w.logCxt.Debugf("Flag node %s as not programmed", name)
+					node.programmedInWireguard = false
+				}
+			}
+		}
+
+		// All updates have been applied. Make sure we delete them after we exit - we will either have applied the deltas,
+		// or we'll need to do a full resync, in either case no need to keep the deltas.  Don't do this immediately because
+		// we may need them to calculate the wireguard config delta.
+		w.peerUpdates = map[string]*peerUpdateData{}
+	}()
+
+	// If necessary ensure the wireguard device is configured. If this errors or if it is not yet oper up then no point
+	// doing anything else.
+	if !w.inSyncLink {
+		w.logCxt.Debug("Ensure wireguard link is created and up")
+		linkUp, err := w.ensureLink(netlinkClient)
+		if netlinkshim.IsNotSupported(err) {
+			// Wireguard is not supported, set everything to "in-sync" since there is not a lot of point doing anything
+			// else. We don't return an error in this case, instead we'll retry every resync period.
+			w.logCxt.Info("Wireguard is not supported - publishing no public key")
+			w.ourPublicKey = &zeroKey
+			w.setAllInSync(true)
+			w.wireguardNotSupported = true
+			return nil
+		} else if err != nil {
+			// Error configuring link, pass up the stack. Close the netlink client as a precaution.
+			w.logCxt.WithError(err).Info("Unable to create wireguard link, retrying...")
+			w.closeNetlinkClient()
+			return ErrUpdateFailed
+		} else if !linkUp {
+			// Wait for oper up notification.
+			w.logCxt.Info("Waiting for wireguard link to come up...")
+			return nil
+		}
+	}
+
+	// Get the wireguard client. This may not always be possible.
+	wireguardClient, err := w.getWireguardClient()
+	if netlinkshim.IsNotSupported(err) {
+		w.logCxt.Info("Wireguard is not supported - send zero-key status")
+		w.ourPublicKey = &zeroKey
+		w.setAllInSync(true)
+		w.wireguardNotSupported = true
+		return nil
+	} else if err != nil {
+		w.logCxt.Errorf("error obtaining wireguard client: %v", err)
```

Comment:
Nit: prefer to do `w.logCxt.WithError(err).Error("Error obtaining wireguard client")`